icu4c/source/common/uloc.cpp

```diff
@@ -1490,7 +1510,10 @@ _canonicalize(const char* localeID,
 
     if (_hasBCP47Extension(localeID)) {
         do {
-            tempBuffer.requestedCapacity = _ConvertBCP47(tmpLocaleID, localeID,
+            localeIDWithHyphens.requestedCapacity = _preprocessLocaleID(localeID, localeIDWithHyphens.getBuffer(), localeIDWithHyphens.getCapacity(), err);
+        } while (localeIDWithHyphens.needToTryAgain(err));
+        do {
+            tempBuffer.requestedCapacity = _ConvertBCP47(tmpLocaleID, U_SUCCESS(*err) ? localeIDWithHyphens.getBuffer() : localeID,
```

Comment:
the problem is then the function would return different result sepending on the memory - but not failing. Both time it will success but return different result. 
I think it is better to return error while the allocation fail, at least you will have a consistent success and consistent failure result and cannot have two different successful results.

Reply:
Yeah, I think you're right.  I'll fix that.